core/src/main/java/org/geysermc/geyser/session/UpstreamSession.java

```diff
@@ -39,30 +40,46 @@
 @RequiredArgsConstructor
 public class UpstreamSession {
     @Getter private final BedrockServerSession session;
+    private final int clientId;
     @Getter @Setter
     private boolean initialized = false;
     private Queue<BedrockPacket> postStartGamePackets = new ArrayDeque<>();
 
     public void sendPacket(@NonNull BedrockPacket packet) {
         if (!isClosed()) {
+            packet.setSenderId(clientId);
             session.sendPacket(packet);
         }
     }
 
     public void sendPacketImmediately(@NonNull BedrockPacket packet) {
         if (!isClosed()) {
+            packet.setSenderId(clientId);
             session.sendPacketImmediately(packet);
         }
     }
 
     public void disconnect(String reason) {
-        session.disconnect(reason);
+        if (isClosed()) {
+            throw new IllegalStateException("Connection has been closed");
+        }
+
+        DisconnectPacket packet = new DisconnectPacket();
+        packet.setSenderId(clientId);
+
+        if (reason == null) {
+            packet.setMessageSkipped(true);
+            reason = "disconnect.disconnected";
+        }
+        packet.setKickMessage(reason);
+        this.sendPacketImmediately(packet);
     }
 
     /**
      * Queue a packet that must be delayed until after login.
      */
     public void queuePostStartGamePacket(BedrockPacket packet) {
+        packet.setClientId(clientId);
```

Comment:
The other methods here use `setSenderId` - I assume this usage here is incorrect?

Reply:
Changed it as suggested - I'm not able to test if splitscreen is still working though. Connecting with a single client still works.